diffoscope: add missing tools

danielfullmer opened the pull request:
###### Motivation for this change
Add a bunch of dependencies for missing tools.

`nix path-info -S` reports the following changes:
diffoscope: 345M -> 356.7M
diffoscope w/ enableBloat: 5.4G -> 6.1G

Also re-enabled the tests and added the dependencies to checkInputs so
the comparator tests are not automatically skipped.


<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->


CC @dezgeg @ma27

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

danielfullmer committed:
diffoscope: add missing tools

`nix path-info -S` reports the following changes:
diffoscope: 345M -> 356.7M
diffoscope w/ enableBloat: 5.4G -> 6.1G

Also re-enabled the tests and added the dependencies to checkInputs so
the comparator tests are not automatically skipped.

Ma27 reviewed (commented):
Tests are failing for me like this:

```
diffoscope> =================================== FAILURES ===================================
diffoscope> _________________________________ test_content _________________________________
diffoscope> differences = [<Difference dir -- dir [<Difference stat {} -- stat {} []>, <Difference text -- text [<Difference stat {} -- stat {} []>]>, <Difference stat {} -- stat {} []>]>]
diffoscope>     def test_content(differences):
diffoscope>         assert differences[0].source1 == 'dir'
diffoscope> >       assert differences[0].details[0].source1 == 'text'
diffoscope> E       AssertionError: assert 'stat {}' == 'text'
diffoscope> E         - stat {}
diffoscope> E         + text
diffoscope> tests/comparators/test_directory.py:71: AssertionError
diffoscope> ============= 1 failed, 344 passed, 98 skipped in 83.97s (0:01:23) =============
diffoscope> builder for '/nix/store/7gnvgnbj2qs3k2792x69k9fss6f4jv2i-diffoscope-144.drv' failed with exit code 1

```

danielfullmer commented:
@Ma27 Hmm, that's strange. It is working for me, and I haven't been able to reproduce. It also seems to be working on the OfBorg builder: https://github.com/NixOS/nixpkgs/pull/89195/checks?check_run_id=722214621

What filesystem are you using? I've tried this on two different machines, using `zfs` and `btrfs`. Could you think of any other impurities that might affect this?

Ma27 commented:
That's interesting. It builds fine on my local machine (NixOS 20.03, Linux 5.6.14, ZFS) and breaks on my remote builder (NixOS 20.03, Linux 5.6.14, ext4 in a KVM).

Unless we find the culprit, I'd suggest to skip the flaky test.

danielfullmer commented:
I was able to reproduce this in a VM with ext4. It turns out this issue was fixed in the most recent version of diffoscope:
https://salsa.debian.org/reproducible-builds/diffoscope/-/issues/74

So, I added a commit bumping to `146`. Built and tested `diffoscope` with and without `enableBloat`.

danielfullmer commented:
Rebased commit against master. I also had to add `apksigner` as a dependency to make the apk tests work with `146` and `enableBloat`.

However, now that I've rebased against master, an indirect dependency is failing to build for me: `multipath-tools`.

Ma27 commented:
@danielfullmer please rebase onto latest master, `multipath-tools` has been fixed in #89367.

danielfullmer commented:
Rebased against master, successfully built and briefly tested with and without `enableBloat`. Should be good to go!
